srcts/src/components/darkModeSwitch.ts

```diff
@@ -0,0 +1,298 @@
+    super.connectedCallback();
+
+    this.themeAttribute =
+      this.getAttribute("theme-attribute") || "data-shinytheme";
+
+    if (typeof this.themeValue === "undefined") {
+      this.themeValue = window.matchMedia("(prefers-color-scheme: dark)")
+        .matches
+        ? "dark"
+        : "light";
+    }
+
+    this.reflectPreference();
+
+    // Sync with system changes
+    window
+      .matchMedia("(prefers-color-scheme: dark)")
+      .addEventListener("change", ({ matches: isDark }) => {
+        this.themeValue = isDark ? "dark" : "light";
+        this.reflectPreference();
+      });
+  }
+
+  getValue(): string {
+    return this.themeValue;
+  }
+
+  render(): ReturnType<LitElement["render"]> {
+    return html`
+      <button
+        title="Switch to ${this.themeValue === "light" ? "dark" : "light"} mode"
+        aria-label="auto"
+        aria-live="polite"
+        data-theme="${this.themeValue}"
+        @click="${this.onClick}"
+      >
+        <svg class="sun-and-moon" aria-hidden="true" viewBox="0 0 24 24">
+          <mask class="moon" id="moon-mask">
+            <rect x="0" y="0" width="100%" height="100%" fill="white" />
+            <circle cx="25" cy="10" r="6" fill="black" />
+          </mask>
+          <circle
+            class="sun"
+            cx="12"
+            cy="12"
+            r="6"
+            mask="url(#moon-mask)"
+            fill="currentColor"
+          />
+          <g class="sun-beams" stroke="currentColor">
+            <line x1="12" y1="1" x2="12" y2="3" />
+            <line x1="12" y1="21" x2="12" y2="23" />
+            <line x1="4.22" y1="4.22" x2="5.64" y2="5.64" />
+            <line x1="18.36" y1="18.36" x2="19.78" y2="19.78" />
+            <line x1="1" y1="12" x2="3" y2="12" />
+            <line x1="21" y1="12" x2="23" y2="12" />
+            <line x1="4.22" y1="19.78" x2="5.64" y2="18.36" />
+            <line x1="18.36" y1="5.64" x2="19.78" y2="4.22" />
+          </g>
+        </svg>
+      </button>
+    `;
+  }
+
+  receiveMessage(el: DarkModeSwitch, data: ToggleMessage): void {
+    if (data.method === "toggle") {
+      if (data.value === "toggle") {
+        data.value = this.themeValue === "light" ? "dark" : "light";
+      }
+      el.setAttribute("theme-value", data.value);
+    }
+  }
+
+  onClick(e: MouseEvent): void {
+    e.stopPropagation();
+    this.themeValue = this.themeValue === "light" ? "dark" : "light";
+  }
+
+  updated(changedProperties: Map<string, any>): void {
+    if (changedProperties.has("themeValue")) {
+      this.reflectPreference();
+      this.onChangeCallback(true);
+    }
+  }
+
+  reflectPreference(): void {
+    document.documentElement.setAttribute(this.themeAttribute, this.themeValue);
+
+    this.shadowRoot
+      ?.querySelector("button")
+      ?.setAttribute("data-theme", this.themeValue);
+
+    this.shadowRoot
+      ?.querySelector("button")
+      ?.setAttribute("aria-label", this.themeValue);
```

Comment:
I'm not completely sure, but I believe `aria-label` should be a constant string, and the state should be reflected in `aria-checked` (or something like that). See https://abbeyperini.medium.com/an-accessible-dark-mode-toggle-in-react-873f7b1495c7

Reply:
Since we're using a `<button>` element here without any internal text, my read (of pages like https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role) is that we should use both `aria-label` and `title` to provide the label. Or, really, we should use `aria-label` but might as well keep `title` for the light-weight hover text.

I'll look into how we should store the current theme value – maybe in `value` instead of `data-theme` - but the linked article used a checkbox input where `aria-checked` makes sense.

Reply:
After reading through https://www.sarasoueidan.com/blog/accessible-icon-buttons/ and reviewing the Bootstrap source for their theme toggle component, I think we should use both `aria-label` and `title` and ensure that the label text includes the current state.

Reply:
Done in 203cd2e4e14b394d3d6c0600c58be60996293510, where now follows `aria-label="Toggle theme ${this.mode}"`; tested with VoiceOver on Safari.

Reply:
I updated the aria label to read `Switch from ${current} to ${other}` in 8dda683c438ef5c3f73081973de2d66b332e1700